automation/entry_scripts/balena-generate-ami.sh

```diff
@@ -0,0 +1,205 @@
+}
+
+deploy_preload_app_to_image() {
+
+    local image=$1
+
+    echo "* Adding the preload app"
+    balena preload \
+      --debug \
+      --app "${BALENA_PRELOAD_APP}" \
+      --commit current \
+      "${image}"
+}
+
+add_ssh_key_to_boot_partition() {
+    local public_key=$1
+
+    echo "* Adding the preload public key"
+    cp "${BOOT_PARTITION}/config.json" /tmp/.config.json
+    jq --arg keys "${public_key}" '. + {os: {sshKeys: [$keys]}}' "${BOOT_PARTITION}/config.json" > /tmp/.config.json
+    mv /tmp/.config.json "${BOOT_PARTITION}/config.json"
+    mount_cleanup
+}
+
+get_value_from_ebs_snapshot_import_task() {
+    local task_id=$1
+    local field=$2
+    aws ec2 describe-import-snapshot-tasks --import-task-ids "${task_id}" | jq -r ".ImportSnapshotTasks[].SnapshotTaskDetail.${field}"
+}
+
+
+create_aws_ebs_snapshot() {
+
+    local img=$1
+    local __retvalue=$2
+
+    local snapshot_id
+    local status
+    local wait_secs=2
+    local secs_waited=0
+    # https://github.com/koalaman/shellcheck/wiki/SC2155#correct-code-1
+    local s3_key
+    # Randomize to lower the chance of parallel builds colliding.
+    s3_key="tmp-$(basename "${img}")-${RANDOM}"
+
+    # Push to s3 and create the AMI
+    echo "* Pushing ${img} to s3://${S3_BUCKET}"
+    aws s3 cp "${img}" "s3://${S3_BUCKET}/preloaded-images/${s3_key}"
+
+    import_task_id=$(aws ec2 import-snapshot \
+      --description "snapshot-${AMI_NAME}" \
+      --disk-container "Description=balenaOs,Format=RAW,UserBucket={S3Bucket=${S3_BUCKET},S3Key=preloaded-images/${s3_key}}" | jq -r .ImportTaskId)
+
+    echo "* Created a AWS import snapshot task with id ${import_task_id}. Waiting for completition... (Timeout: $IMPORT_SNAPSHOT_TIMEOUT_MINS mins)"
+    while true; do
+        status="$(get_value_from_ebs_snapshot_import_task "${import_task_id}" Status)"
+        [[ "$status" == "completed" ]] && break
+        [[ "$status" == "deleting" ]]  && \
+            error_msg="$(get_value_from_ebs_snapshot_import_task "${import_task_id}" StatusMessage)" && \
+            echo "ERROR: Error on import task id ${import_task_id}: '${error_msg}'" && exit 1
+
+        sleep $wait_secs
+        secs_waited=$((secs_waited + wait_secs))
+        mins_elapsed=$((secs_waited / 60))
+
+        # Show progress every 2 mins (120 secs)
+        [[ $((secs_waited % 120)) == 0 ]] && echo "-> Mins elapsed: $mins_elapsed. Progress: $(get_value_from_ebs_snapshot_import_task "${import_task_id}" Progress)%"
+        [[ "$mins_elapsed" -ge "$IMPORT_SNAPSHOT_TIMEOUT_MINS" ]] && echo "ERROR: Timeout on import snapshot taksk id ${import_task_id}" && exit 1
+    done
+
+    snapshot_id=$(aws ec2 describe-import-snapshot-tasks --import-task-ids "${import_task_id}" | jq -r '.ImportSnapshotTasks[].SnapshotTaskDetail.SnapshotId')
+    echo "* AWS import snapshot task complete. SnapshotId: ${snapshot_id}"
+
+    echo "* Removing img from S3..."
+    aws s3 rm "s3://${S3_BUCKET}/preloaded-images/${s3_key}"
```

Comment:
@mtoman probably need to put this into a trap handler, otherwise failed jobs may leave around cruft in our S3 bucket(s).